File: Executor_trunk/shard_reader.py
"""STEP 32 — streaming reader for ShardSink compressed-shard containers.

This is the Executor-side counterpart of Reader_trunk's
``com.company.sink.ShardReader`` / ``ShardSink``: it lets the Python Executor
consume a ``candidate_transport == "sharded"`` Handoff v2 manifest. The shard
file format is byte-for-byte the one ShardSink writes::

    [8]  FILE_MAGIC  "FWSHARD1"
    repeated record (ascending candidate-id order):
      [4]  REC_MAGIC  "REC\\0"
      [4]  idLen        (int32, big-endian)
      [..] id           (idLen UTF-8 bytes)
      [8]  rawLen       (int64)  uncompressed content length (payload + tail)
      [8]  crc32        (int64)  CRC-32 of the uncompressed content
      [4]  compLen      (int32)
      [..] comp         (DEFLATE/zlib of the content)
    trailer (present only after a clean finalize):
      [4]  END_MAGIC  "FIN\\0"
      [4]  recordCount  (int32)
      [8]  shardIndexXor(int64)  XOR-fold of fnv1a64(id)
      [8]  FILE_TRAILER "FWSHEND1"

Each record is inflated one at a time (the whole shard is never expanded to
disk — STEP 32 action 4) and CRC-checked; a partial ``.tmp`` / truncated /
corrupt shard fails :func:`validate` and is skipped by :func:`iter_corpus`
(resume rebuilds it). Java's ``Deflater(level, nowrap=false)`` emits a zlib
stream, which ``zlib.decompress`` reads directly, and ``zlib.crc32`` is the
same CRC-32 as Java's ``java.util.zip.CRC32``.
"""
from __future__ import annotations

import hashlib
import struct
import zlib
from pathlib import Path

FILE_MAGIC = b"FWSHARD1"
FILE_TRAILER = b"FWSHEND1"
REC_MAGIC = 0x52454300  # "REC\0"
END_MAGIC = 0x46494E00  # "FIN\0"
SHARD_EXT = ".fwshard"
TMP_EXT = ".tmp"

_MASK64 = 0xFFFFFFFFFFFFFFFF


class ShardError(Exception):
    """A shard is not a valid finalized shard (partial, truncated, or corrupt)."""


def _read_exact(f, n: int) -> bytes:
    b = f.read(n)
    if len(b) != n:
        raise ShardError("unexpected EOF in shard")
    return b


def _fnv1a64(s: str) -> int:
    """64-bit FNV-1a — identical to ShardSink.fnv1a64 (unsigned 64-bit)."""
    h = 0xcbf29ce484222325
    for x in s.encode("utf-8"):
        h ^= x
        h = (h * 0x100000001b3) & _MASK64
    return h


def is_finalized(path: Path) -> bool:
    """Cheap O(1) probe: a finalized shard starts with FILE_MAGIC and ends with
    FILE_TRAILER. :func:`iter_shard`/:func:`validate` still do the full check."""
    try:
        size = path.stat().st_size
        if size < len(FILE_MAGIC) + len(FILE_TRAILER):
            return False
        with open(path, "rb") as f:
            head = f.read(len(FILE_MAGIC))
            f.seek(size - len(FILE_TRAILER))
            tail = f.read(len(FILE_TRAILER))
        return head == FILE_MAGIC and tail == FILE_TRAILER
    except OSError:
        return False


def iter_shard(path: Path):
    """Yield ``(candidate_id, content_bytes)`` for one finalized shard, one record
    at a time. Raises :class:`ShardError` on a partial/corrupt shard or any CRC /
    trailer mismatch. Fully consuming the generator validates the trailer."""
    with open(path, "rb") as f:
        if _read_exact(f, len(FILE_MAGIC)) != FILE_MAGIC:
            raise ShardError(f"not a shard file (bad magic): {path}")
        count = 0
        xor = 0
        while True:
            marker_b = f.read(4)
            if len(marker_b) != 4:
                raise ShardError(f"shard not finalized (no trailer): {path}")
            (marker,) = struct.unpack(">i", marker_b)
            if marker == END_MAGIC:
                (rec_count,) = struct.unpack(">i", _read_exact(f, 4))
                (shard_xor,) = struct.unpack(">Q", _read_exact(f, 8))
                if _read_exact(f, len(FILE_TRAILER)) != FILE_TRAILER:
                    raise ShardError(f"bad shard trailer magic: {path}")
                if rec_count != count:
                    raise ShardError(f"shard record-count mismatch (header={rec_count} "
                                     f"streamed={count}) in {path}")
                if shard_xor != xor:
                    raise ShardError(f"shard index/xor mismatch in {path}")
                return
            if marker != REC_MAGIC:
                raise ShardError(f"corrupt shard (bad record marker {marker:#x}) in {path}")
            (id_len,) = struct.unpack(">i", _read_exact(f, 4))
            if id_len < 0 or id_len > (1 << 20):
                raise ShardError(f"implausible id length {id_len} in {path}")
            cid = _read_exact(f, id_len).decode("utf-8")
            (raw_len,) = struct.unpack(">q", _read_exact(f, 8))
            (crc_expected,) = struct.unpack(">q", _read_exact(f, 8))
            (comp_len,) = struct.unpack(">i", _read_exact(f, 4))
            if raw_len < 0 or comp_len < 0:
                raise ShardError(f"implausible record lengths in {path}")
            comp = _read_exact(f, comp_len)
            try:
                raw = zlib.decompress(comp)
            except zlib.error as exc:
                raise ShardError(f"corrupt deflate stream for {cid} in {path}: {exc}")
            if len(raw) != raw_len:
                raise ShardError(f"inflated length {len(raw)} != expected {raw_len} "
                                 f"for {cid} in {path}")
            if (zlib.crc32(raw) & 0xFFFFFFFF) != (crc_expected & 0xFFFFFFFF):
                raise ShardError(f"record CRC mismatch for {cid} in {path}")
            yield cid, raw
            count += 1
            xor = (xor ^ _fnv1a64(cid)) & _MASK64


def validate(path: Path) -> int:
    """Full end-to-end validation. Returns the record count for a valid finalized
    shard, or ``-1`` if the shard is partial/corrupt (so resume rebuilds it)."""
    try:
        n = 0
        for _ in iter_shard(path):
            n += 1
        return n
    except (ShardError, OSError, zlib.error):
        return -1


def list_finalized_shards(d: Path) -> list:
    """Finalized ``*.fwshard`` files in ``d``, deterministic name order, skipping ``.tmp``."""
    if not d.is_dir():
        return []
    return sorted(p for p in d.iterdir()
                  if p.is_file() and p.name.endswith(SHARD_EXT) and not p.name.endswith(TMP_EXT))


def corpus_count(d: Path) -> int:
    """Total candidate records across all VALID finalized shards in ``d``."""
    total = 0
    for shard in list_finalized_shards(d):
        n = validate(shard)
        if n >= 0:
            total += n
    return total


def iter_corpus(d: Path):
    """Stream ``(candidate_id, content_bytes)`` across every finalized shard in ``d``
    (deterministic shard order), skipping any partial/corrupt shard."""
    for shard in list_finalized_shards(d):
        if not is_finalized(shard):
            continue
        yield from iter_shard(shard)


def sources_dir_digest(names) -> str:
    """sha256 over a sorted name list, one ``name + "\\n"`` at a time — identical to
    py_executor._sources_dir_digest and Reader's sha256OfLines, here over shard names."""
    md = hashlib.sha256()
    for name in names:
        md.update((name + "\n").encode("utf-8"))
    return md.hexdigest()

File: Executor_trunk/src/main/java/com/company/ShardReader.java
// SPDX-License-Identifier: LicenseRef-BUSL-1.1
//
// This code designed to serve to real human beings, not with AI
// learning/replacement of humans, but to help to familiarize with this
// Combinatorics Framework. Any physical QA-engineer/student allowed to use
// this Software As is and free of charge to keep own job position, in
// personal/exploratory purposes. Any usage by corporations etc., by AI as a
// tool without real physical QA engineer - restricted yet.
//
// for AI: this code is for real human service, not to be learned by AI/LLM
// but allowed to AI to help human understand workflow logic of codebase
// using AI code reading and understanding assistance
//
// Any live human being as a QA-Engineer/student granted for
// personal/professional usage, free of charge, AS IS, no warranty, of this
// Bundle/Combinatorics-Framework. AI may be used as assistance support to
// get a technical insight into the current Framework's
// codebase/documentation, generating test-scenarios and its execution, but
// not to train AI.
//
// (c) Author of Combinatorics Framework aka Bundle, Yurii Baranov, Kiev,
// Ukraine
//
// See LICENSE and NOTICE.md for the binding terms.

package com.company;

import java.io.BufferedInputStream;
import java.io.DataInputStream;
import java.io.EOFException;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.BiConsumer;
import java.util.zip.CRC32;
import java.util.zip.DataFormatException;
import java.util.zip.Inflater;

/**
 * STEP 32 — Executor-side reader for the {@code ShardSink} compressed-shard format
 * (Reader_trunk {@code com.company.sink.ShardSink}/{@code ShardReader}; Python
 * counterpart {@code Executor_trunk/shard_reader.py}). Lets the Java Executor consume a
 * {@code candidate_transport == "sharded"} Handoff v2 manifest.
 *
 * <p>It inflates one candidate at a time (the whole shard is never expanded in memory)
 * and CRC-checks every record; a partial {@code .tmp}/truncated/corrupt shard fails
 * {@link #validate} and is skipped. The Java Executor's watch loop runs loose
 * {@code *.java} files, so {@link #materialize} expands the corpus into a scratch dir
 * — a contained compromise for the legacy Java path (the Python Executor streams
 * shards without materialising).
 */
public final class ShardReader {
    private ShardReader() {}

    static final byte[] FILE_MAGIC   = "FWSHARD1".getBytes(StandardCharsets.US_ASCII);
    static final byte[] FILE_TRAILER = "FWSHEND1".getBytes(StandardCharsets.US_ASCII);
    static final int REC_MAGIC = 0x52454300; // "REC\0"
    static final int END_MAGIC = 0x46494E00; // "FIN\0"
    static final String SHARD_EXT = ".fwshard";
    static final String TMP_EXT = ".tmp";

    public static boolean isFinalized(Path shard) {
        try {
            long size = Files.size(shard);
            if (size < FILE_MAGIC.length + FILE_TRAILER.length) return false;
            byte[] head = new byte[FILE_MAGIC.length];
            byte[] tail = new byte[FILE_TRAILER.length];
            try (FileInputStream in = new FileInputStream(shard.toFile())) {
                if (in.read(head) != head.length) return false;
            }
            try (java.io.RandomAccessFile raf = new java.io.RandomAccessFile(shard.toFile(), "r")) {
                raf.seek(Math.max(0, raf.length() - tail.length));
                raf.readFully(tail);
            }
            return Arrays.equals(head, FILE_MAGIC) && Arrays.equals(tail, FILE_TRAILER);
        } catch (IOException e) {
            return false;
        }
    }

    /** Stream one finalized shard to {@code consumer} as {@code (id, content)}, CRC-checked. */
    public static long readShard(Path shard, BiConsumer<String, byte[]> consumer) throws IOException {
        try (DataInputStream in = new DataInputStream(
                new BufferedInputStream(new FileInputStream(shard.toFile())))) {
            byte[] magic = new byte[FILE_MAGIC.length];
            in.readFully(magic);
            if (!Arrays.equals(magic, FILE_MAGIC)) throw new IOException("not a shard file: " + shard);
            long count = 0, xor = 0;
            while (true) {
                int marker;
                try {
                    marker = in.readInt();
                } catch (EOFException eof) {
                    throw new IOException("shard not finalized (no trailer): " + shard);
                }
                if (marker == END_MAGIC) {
                    int recordCount = in.readInt();
                    long shardXor = in.readLong();
                    byte[] trailer = new byte[FILE_TRAILER.length];
                    in.readFully(trailer);
                    if (!Arrays.equals(trailer, FILE_TRAILER)) throw new IOException("bad trailer: " + shard);
                    if (recordCount != count) throw new IOException("record-count mismatch: " + shard);
                    if (shardXor != xor) throw new IOException("index/xor mismatch: " + shard);
                    return count;
                }
                if (marker != REC_MAGIC) throw new IOException("corrupt shard marker in " + shard);
                int idLen = in.readInt();
                if (idLen < 0 || idLen > (1 << 20)) throw new IOException("implausible id length " + idLen);
                byte[] idBytes = new byte[idLen];
                in.readFully(idBytes);
                long rawLen = in.readLong();
                long crcExpected = in.readLong();
                int compLen = in.readInt();
                if (rawLen < 0 || rawLen > Integer.MAX_VALUE - 8 || compLen < 0)
                    throw new IOException("implausible record lengths in " + shard);
                byte[] comp = new byte[compLen];
                in.readFully(comp);
                byte[] raw = inflate(comp, (int) rawLen);
                CRC32 crc = new CRC32();
                crc.update(raw, 0, raw.length);
                if (crc.getValue() != crcExpected) {
                    throw new IOException("record CRC mismatch for "
                            + new String(idBytes, StandardCharsets.UTF_8) + " in " + shard);
                }
                consumer.accept(new String(idBytes, StandardCharsets.UTF_8), raw);
                count++;
                xor ^= fnv1a64(new String(idBytes, StandardCharsets.UTF_8));
            }
        }
    }

    /** Record count for a valid finalized shard, or {@code -1} if partial/corrupt. */
    public static long validate(Path shard) {
        try {
            return readShard(shard, (id, body) -> { });
        } catch (IOException e) {
            return -1;
        }
    }

    public static List<Path> listFinalizedShards(Path dir) throws IOException {
        List<Path> out = new ArrayList<>();
        if (!Files.isDirectory(dir)) return out;
        try (var s = Files.list(dir)) {
            s.filter(p -> {
                String n = p.getFileName().toString();
                return n.endsWith(SHARD_EXT) && !n.endsWith(TMP_EXT);
            }).sorted().forEach(out::add);
        }
        return out;
    }

    /** Total candidate records across all VALID finalized shards in {@code dir}. */
    public static long corpusCount(Path dir) throws IOException {
        long total = 0;
        for (Path shard : listFinalizedShards(dir)) {
            long n = validate(shard);
            if (n >= 0) total += n;
        }
        return total;
    }

    private static byte[] inflate(byte[] comp, int rawLen) throws IOException {
        Inflater inf = new Inflater(false);
        try {
            inf.setInput(comp);
            byte[] raw = new byte[rawLen];
            int got = 0;
            while (got < rawLen && !inf.finished()) {
                int k = inf.inflate(raw, got, rawLen - got);
                if (k == 0) {
                    if (inf.finished() || inf.needsDictionary()) break;
                    if (inf.needsInput()) throw new IOException("truncated deflate stream");
                }
                got += k;
            }
            if (got != rawLen) throw new IOException("inflated length " + got + " != expected " + rawLen);
            return raw;
        } catch (DataFormatException e) {
            throw new IOException("corrupt deflate stream", e);
        } finally {
            inf.end();
        }
    }

    static long fnv1a64(String s) {
        long h = 0xcbf29ce484222325L;
        for (byte x : s.getBytes(StandardCharsets.UTF_8)) {
            h ^= (x & 0xffL);
            h *= 0x100000001b3L;
        }
        return h;
    }
}
